app/main.py

```diff
@@ -1,6 +1,44 @@
 from typing import Callable
+from functools import wraps
 
 
 def cache(func: Callable) -> Callable:
-    # Write your code here
-    pass
+    cache_storage = {}
+
+    @wraps(func)
+    def wrapper(*args) -> Callable:
+        if args in cache_storage:
+            print("Getting from cache")
+            return cache_storage[args]
+        else:
+            print("Calculating new result")
+            result = func(*args)
+            cache_storage[args] = result
+            return result
+
+    return wrapper
+
+
+@cache
+def long_time_func(base: int, exponent: int, mod_factor: int) -> int:
+    return (base ** exponent ** mod_factor) % (base * mod_factor)
+
+
+@cache
+def long_time_func_2(n_tuple: tuple, power: int) -> int:
+    return [number ** power for number in n_tuple]
```

Comment:
The `long_time_func_2` function returns a list, which is not hashable and cannot be used as a cache key. Consider converting the list to a tuple before returning, or modify the caching mechanism to handle non-hashable return types.